libfwupdplugin/fu-progress.c

```diff
@@ -0,0 +1,774 @@
+
+static guint signals[SIGNAL_LAST] = {0};
+
+G_DEFINE_TYPE_WITH_PRIVATE(FuProgress, fu_progress, G_TYPE_OBJECT)
+#define GET_PRIVATE(o) (fu_progress_get_instance_private(o))
+
+/**
+ * fu_progress_get_id:
+ * @self: a #FuProgress
+ *
+ * Return the id of the progress, which is normally set by the caller.
+ *
+ * Returns: progress ID
+ *
+ * Since: 1.7.0
+ **/
+const gchar *
+fu_progress_get_id(FuProgress *self)
+{
+	FuProgressPrivate *priv = GET_PRIVATE(self);
+	g_return_val_if_fail(FU_IS_PROGRESS(self), NULL);
+	return priv->id;
+}
+
+/**
+ * fu_progress_set_id:
+ * @self: a #FuProgress
+ * @id: progress ID, normally `G_STRLOC`
+ *
+ * Sets the id of the progress.
+ *
+ * Since: 1.7.0
+ **/
+void
+fu_progress_set_id(FuProgress *self, const gchar *id)
+{
+	FuProgressPrivate *priv = GET_PRIVATE(self);
+	g_return_if_fail(FU_IS_PROGRESS(self));
+	g_return_if_fail(id != NULL);
+
+	/* not changed */
+	if (g_strcmp0(priv->id, id) == 0)
+		return;
+
+	/* set id */
+	g_free(priv->id);
+	priv->id = g_strdup(id);
+}
+
+/**
+ * fu_progress_get_status:
+ * @self: a #FuProgress
+ *
+ * Return the status of the progress, which is normally indirectly by fu_progress_add_step().
+ *
+ * Returns: status
+ *
+ * Since: 1.7.0
+ **/
+FwupdStatus
+fu_progress_get_status(FuProgress *self)
+{
+	FuProgressPrivate *priv = GET_PRIVATE(self);
+	g_return_val_if_fail(FU_IS_PROGRESS(self), FWUPD_STATUS_UNKNOWN);
+	return priv->status;
+}
+
+/**
+ * fu_progress_set_status:
+ * @self: a #FuProgress
+ * @status: device status
+ *
+ * Sets the status of the progress.
+ *
+ * Since: 1.7.0
+ **/
+void
+fu_progress_set_status(FuProgress *self, FwupdStatus status)
+{
+	FuProgressPrivate *priv = GET_PRIVATE(self);
+	g_return_if_fail(FU_IS_PROGRESS(self));
+
+	/* not changed */
+	if (priv->status == status)
+		return;
+
+	/* save */
+	priv->status = status;
+	g_signal_emit(self, signals[SIGNAL_STATUS_CHANGED], 0, status);
+}
+
+/**
+ * fu_progress_get_percentage:
+ * @self: a #FuProgress
+ *
+ * Get the last set progress percentage.
+ *
+ * Return value: The percentage value, or %G_MAXUINT for error
+ *
+ * Since: 1.7.0
+ **/
+guint
+fu_progress_get_percentage(FuProgress *self)
+{
+	FuProgressPrivate *priv = GET_PRIVATE(self);
+	g_return_val_if_fail(FU_IS_PROGRESS(self), G_MAXUINT);
+	return priv->percentage;
+}
+
+static void
+fu_progress_build_parent_chain(FuProgress *self, GString *str, guint level)
+{
+	FuProgressPrivate *priv = GET_PRIVATE(self);
+	if (priv->parent != NULL)
+		fu_progress_build_parent_chain(priv->parent, str, level + 1);
+	g_string_append_printf(str,
+			       "%u) %s (%u/%u)\n",
+			       level,
+			       priv->id,
+			       priv->step_now,
+			       priv->step_max);
+}
+
+/**
+ * fu_progress_set_percentage:
+ * @self: a #FuProgress
+ * @percentage: value between 0% and 100%
+ *
+ * Sets the progress percentage complete.
+ *
+ * NOTE: this must be above what was previously set, or it will be rejected.
```

Comment:
Rather than a warning level rejection, how about making it an error and this return a boolean?

Reply:
I also pondered this. I think in the case where the plugin author got the FuProgress stuff wrong we still want to update the firmware on the device, i.e. it's better we do the #1 purpose of the tool and nag on the console than complain to the user that can't actually do anything with the failure.